docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,329 @@
+
+  `Fetch`  will not combine object storage requests.
+
+  It will split big range request to multiple small requests if needed.
+
+   `Fetch` internals:
+
+  * ensure `end` is not greater than `f.size`
+
+  * convert range to page ids
+* get new page ids (not in f.pages)
+  * if have not new page ids
+  * return a slice of mmaped file
+  * split new page ids to multiple partitions
+  * convert partition page ids to range and fetch concurrently
+  * return a slice of mmaped file
+  
+* Prefetch: 
+
+  `Prefetch` will combine small object storage requests. it returns a channel to pending callers, the channel is used to notify the result of prefetch.
+
+  Callers need to check value from the channel to know if the prefetch was successful. If prefetch done, then callers can call the `Range` method to retrieve bytes.
+
+  `Prefetch` also returns an error if the error can be found quickly (e.g. `end` is greater than `f.size`).
+
+  `Prefetch` internals:
+
+  - ensure `end` is not great than `f.size`
+
+  - convert ranges to page ids
+- get new page ids (not in f.pages)
+  - if have not new page ids (have been fetched)
+  - return a slice of mmaped file (f.data[start:end])
+  
+- add new page into `f.pendingPages`
+  - increase `f.pendingReaders`
+  - return `f.pendingChan`
+* Range: 
+
+  `Range` returns a slice of mmaped file (`f.data[start:end]`). 
+
+  It will not try to fetch data from obj store, and return an error if some pages in the range have not been fetched.
+
+  `Range` internals:
+
+  - convert range to page ids  
+  - if all page ids have been fetched
+    - return a slice of mmaped file
+
+  * otherwise return an error
+
+* CloseMmap: `CloseMmap` unmmaps `f.data` .
+
+* CloseFile: `CloseFile` closes  `f.file`. 
+
+* Close: `Close` closes `f.file`, unmmaps `f.data` and set `f.pages` to nil.
+
+### Range to pages
+
+Convert a range to pages:
+
+```
+const pageSize = 4 * 1024  // 4KiB
+
+minPageID := math.Floor( float(start) / float(pageSize) )
+maxPageID := math.Ceil( float(end) / float(pageSize) )
+
+pages := roaring.NewBitmap()
+pages.AddRange(minPageID, maxPageID)
+```
+
+### Combine queries
+
+```
+type BucketFile struct {
+    // ...
+    pendingDuration time.Duration   // duration to fetch pending pages
+    pendingPages    *roaring.Bitmap // pending pages to be fetched
+    pendingReaders  int             // record pending callers
+    pendingChan     chan error      // chan to notify callers the result of prefetch	
+}
+
+func (f *BucketFile) Prefetch(ranges []*Range) (chan error, error) {...}
+```
+
+BucketFile appends new pages to `pendingPages`, waits for next  time to fetch. So that we can combine multiple queries to ones.
+
+When fetching combined  pages,
+if an error occurs, BucketFile will send the error to callers via the `pendingChan` channel,
```

Comment:
Error handling is really a low-level implementation detail and should be not part of the design as it does not add much to the topic. It's enough to say - there must be error handling for failed prefetches. And if single fetch is retried or crash whole request? Those questions should be covered in design, not what go channels to use, I think (: